src/functions/module_filter.ts

```diff
@@ -1,9 +1,95 @@
+import { cellsSortingCriteria } from "../helpers/sort";
 import { _t } from "../translation";
-import { AddFunctionDescription, Arg, isMatrix, Matrix, Maybe, ValueAndFormat } from "../types";
+import {
+  AddFunctionDescription,
+  Arg,
+  CellValue,
+  CellValueType,
+  isMatrix,
+  Locale,
+  Matrix,
+  Maybe,
+  ValueAndFormat,
+} from "../types";
 import { NotAvailableError } from "../types/errors";
 import { arg } from "./arguments";
 import { assertSameDimensions, assertSingleColOrRow } from "./helper_assert";
-import { assert, matrixMap, toBoolean, toMatrix, transposeMatrix } from "./helpers";
+import { toScalar } from "./helper_matrices";
+import {
+  assert,
+  matrixMap,
+  toBoolean,
+  toMatrix,
+  toNumber,
+  toString,
+  transposeMatrix,
+} from "./helpers";
+
+function sortMatrix(
+  _range: Matrix<ValueAndFormat>,
+  locale: Locale,
+  ...values: (ValueAndFormat | Matrix<ValueAndFormat>)[]
+): Matrix<ValueAndFormat> {
+  const criteriaIndex: (0 | 1)[] = [];
+  const sortColumns: Matrix<CellValue> = [];
+  const nRows = _range.length;
+  for (let i = 0; i < values.length; i += 2) {
+    criteriaIndex.push(toBoolean(toScalar(values[i + 1]).value) ? 1 : 0);
+    const sortColumn = values[i];
+    if (isMatrix(sortColumn) && (sortColumn.length > 1 || sortColumn[0].length > 1)) {
+      assert(
+        () => sortColumn.length === 1 && sortColumn[0].length === nRows,
+        _t(
+          `Wrong size of Argument[%s]. Expected a range of size 1x%s. Got %sx%s.`,
+          i + 1,
+          nRows,
+          sortColumn.length,
+          sortColumn[0].length
+        )
+      );
+      sortColumns.push(sortColumn.flat().map((v) => v.value));
+    } else {
+      const colIndex = toNumber(toScalar(sortColumn).value, locale);
+      if (colIndex < 1 || colIndex > _range[0].length) {
+        return _range;
+      }
+      sortColumns.push(_range.map((row) => row[colIndex - 1].value));
+    }
+  }
+  if (sortColumns.length === 0) {
+    for (let i = 0; i < _range[0].length; i++) {
+      sortColumns.push(_range.map((row) => row[i].value));
+      criteriaIndex.push(1);
+    }
+  }
+  const sortingCriteria = [cellsSortingCriteria(false), cellsSortingCriteria(true)];
+  const indexes: number[] = [];
+  for (let row = 0; row < _range.length; row++) {
+    indexes.push(row);
+  }
+  indexes.sort((a, b) => {
+    for (const [i, sortColumn] of sortColumns.entries()) {
+      const left = sortColumn[a];
+      const right = sortColumn[b];
+      // TODO: This is very very moche !
+      const leftCell = {
+        value: left,
+        type: left === null ? CellValueType.empty : typeof left === "string" ? "text" : typeof left,
```

Comment:
check if the empty string should not be considered as `empty`.

Reply:
It shouldn't :)